subcerts.go

```diff
@@ -0,0 +1,533 @@
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+)
+
+const (
+	dcMaxTTLSeconds   = 60 * 60 * 24 * 7 // 7 days
+	dcMaxTTL          = time.Duration(dcMaxTTLSeconds * time.Second)
+	dcMaxPublicKeyLen = 1 << 16 // Bytes
+	dcMaxSignatureLen = 1 << 16 // Bytes
+)
+
+var errNoDelegationUsage = errors.New("certificate not authorized for delegation")
+
+// delegationUsageId is the DelegationUsage X.509 extension OID
+//
+// TODO(any) Replace with the real OID. This value is temporary.
+var delegationUsageId = asn1.ObjectIdentifier{2, 5, 29, 99}
+
+// CreateDelegationUsagePKIXExtension returns a pkix.Extension that every delegation
+// certificate must have.
+func CreateDelegationUsagePKIXExtension() *pkix.Extension {
+	return &pkix.Extension{
+		Id:       delegationUsageId,
+		Critical: false,
+		Value:    nil,
+	}
+}
+
+// canDelegate returns true if a certificate can be used for delegated
+// credentials.
+func canDelegate(cert *x509.Certificate) bool {
+	// Check that the digitalSignature key usage is set.
+	if (cert.KeyUsage & x509.KeyUsageDigitalSignature) == 0 {
+		return false
+	}
+
+	// Check that the certificate has the DelegationUsage extension and that
+	// it's non-critical (per the spec).
+	for _, extension := range cert.Extensions {
+		if extension.Id.Equal(delegationUsageId) {
+			return true
+		}
+	}
+	return false
+}
+
+// This structure stores the public components of a credential.
+type credential struct {
+	validTime time.Duration
+	publicKey crypto.PublicKey
+	scheme    SignatureScheme
+}
+
+// marshalSubjectPublicKeyInfo returns a DER encoded SubjectPublicKeyInfo structure
+// (as defined in the X.509 standard) for the credential.
+func (cred *credential) marshalSubjectPublicKeyInfo() ([]byte, error) {
+	switch cred.scheme {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512:
+
+		serializedPublicKey, err := x509.MarshalPKIXPublicKey(cred.publicKey)
+		if err != nil {
+			return nil, err
+		}
+		return serializedPublicKey, nil
+
+	default:
+		return nil, fmt.Errorf(
+			"unsupported signature scheme: 0x%04x", cred.scheme)
+	}
+}
+
+// unmarshalSubjectPublicKeyInfo parses a DER encoded SubjectPublicKeyINfo
+// structure into a public key and its corresponding algorithm.
+func unmarshalSubjectPublicKeyInfo(
+	serialized []byte) (crypto.PublicKey, SignatureScheme, error) {
+
+	publicKey, err := x509.ParsePKIXPublicKey(serialized)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	switch pk := publicKey.(type) {
+	case *ecdsa.PublicKey:
+
+		curveName := pk.Curve.Params().Name
+		if curveName == "P-256" {
+			return pk, ECDSAWithP256AndSHA256, nil
+		} else if curveName == "P-384" {
+			return pk, ECDSAWithP384AndSHA384, nil
+		} else if curveName == "P-521" {
+			return pk, ECDSAWithP521AndSHA512, nil
+		} else {
+			return nil, 0, fmt.Errorf("curve %s s not supported", curveName)
+		}
+
+	default:
+		return nil, 0, fmt.Errorf(
+			"unsupported delgation key type: %s", reflect.TypeOf(pk))
```

Comment:
unwrap, fmt.Errorf("... %T", pk) so you don't need to import reflect

Reply:
Done.